pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -180,39 +177,33 @@ func (a *oauthTokenTimeoutValidator) flush(flushHorizon time.Time) {
 
 	glog.V(5).Infof("Flushing tokens timing out before %v", flushHorizon)
 
-	for {
-		item := a.tree.Min()
-		if item == nil {
-			// out of items
-			break
-		}
+	for item := a.tree.Min(); item != nil; item = a.tree.Min() {
 		tdr := item.(*tokenDataRef)
-		td := a.data[tdr.name]
-		if item.(*tokenDataRef).timeout.Before(flushHorizon) {
-			delta := a.clientTimeout(td.client)
-			var newtimeout int32
-			if delta > 0 {
-				newtimeout = int32((td.seen.Sub(td.creation) + delta) / time.Second)
-			} else {
-				newtimeout = 0
-			}
-			patch := []byte(fmt.Sprintf(`[{"op": "test", "path": "/timeoutsIn", "value": %d}, {"op": "replace", "path": "/timeoutsIn", "value": %d}]`, td.timeout, newtimeout))
-			_, err := a.tokens.Patch(td.name, ktypes.JSONPatchType, patch)
-			if err != nil {
-				glog.V(5).Infof("Token timeout was not updated: %v", err)
-			}
-			a.remove(td, tdr)
-			flushedTokens++
-		} else {
+		if tdr.timeout.After(flushHorizon) {
 			// out of items within the flush Horizon
 			break
 		}
+
+		td := a.data[tdr.name]
+		delta := a.clientTimeout(td.token.ClientName)
+		var newtimeout int32
+		if delta > 0 {
+			newtimeout = int32((td.seen.Sub(td.token.CreationTimestamp.Time) + delta) / time.Second)
+		}
+		patch := []byte(fmt.Sprintf(`[{"op": "test", "path": "/timeoutsIn", "value": %d}, {"op": "replace", "path": "/timeoutsIn", "value": %d}]`, td.token.TimeoutsIn, newtimeout))
+		_, err := a.tokens.Patch(td.token.Name, ktypes.JSONPatchType, patch)
+		if err != nil {
+			glog.V(5).Infof("Token timeout was not updated: %v", err)
+			// TODO can we not remove and try later?
```

Comment:
no it genrally means another master has already updated the token with a newer timeout value or the token has been deleted. If the user keeps using a new token will hopefully come in with the right values.